[DNM] Reference Tracking on Live Again Again

LemonInTheDark opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Enables reftracking on live + any fixes/cleanup I do along the way
I'll be closing this pr and opening one that just contains the fixes scraped from this round of testing once I'm done.

## Why It's Good For The Game

Causes a good amount of lag, bit of stutter
Hope is'll find harddels that can be fixed, so pain now for gain later

Something something muh atmos cpu time idk
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

LemonInTheDark committed:
Fixes robot customor sourced sign harddels

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into here-we-go

LemonInTheDark committed:
Adds a define that enables live reference tracking, making testing less of a headache, and hopefully enabling others to do this more easily

LemonInTheDark committed:
Catches bad machine addition

LemonInTheDark committed:
I'm a dumbass, weakref properly

LemonInTheDark committed:
Minor improvement to ref tracking define errors

LemonInTheDark committed:
Enables reference tracking

Mothblocks commented:
Something you might be interested is the fact that we (key holders) can run defines in `server_side_modifications.dm`.

If you could make this a PR that can be easily flipped on by a define, you could potentially just contact me to flip it on for a server.

LemonInTheDark committed:
Fixes aura healing alert harddels, why are you holding a ref you don't need, did you test this? God

LemonInTheDark committed:
Converts isproperdatum to a define, not sure why it wouldn't be one

LemonInTheDark committed:
Uses signals to prevent actions from hanging a ref to their target, I am very sad

LemonInTheDark committed:
Fixes a borg screen object harddel, we never nulled this. Why? No idea

LemonInTheDark committed:
Memories were passing atoms into extra_info, to be flattened into strings later. This hangs a ton of refs (I suspect this is what kills surgeries) and we only use the refs to flatten them, so why not just do that to start with?

LemonInTheDark committed:
Fixes an uplink component harddel, this is why god invented weakrefs

LemonInTheDark committed:
Fixes the orbit menu hanging refs to deleted points of interest HHHHHHHHHHHHHHH

LemonInTheDark committed:
Whoops, this needs to actually compile now that I think about it

LemonInTheDark committed:
See above\

LemonInTheDark committed:
Oh. Shoulda thought about that one huh

LemonInTheDark committed:
Merge branch 'here-we-go' into here-we-suffer

LemonInTheDark commented:
> Something you might be interested is the fact that we (key holders) can run defines in `server_side_modifications.dm`.
> 
> If you could make this a PR that can be easily flipped on by a define, you could potentially just contact me to flip it on for a server.

Totally coincidentally, this is already the plan
I don't like the idea of needing me to exist to do reftracking on live, so I've made a define that toggles it + logging + crash prevention.

Outside of this, the run's going well, caught quite a few things so far

LemonInTheDark committed:
Remove can loop apperantly, let's just uh

LemonInTheDark committed:
Merge branch 'here-we-go' into here-we-suffer

LemonInTheDark committed:
Fixes mech actions holding a ref to their mech, makes removing them actually qdel the actions, which is part of the reason this happen in the first place

LemonInTheDark committed:
Turns the recharging mech var into a weakref, fixing harddels. also cleans things up a smidge

LemonInTheDark committed:
Fixes bluespace vendor sourced tank harddels

LemonInTheDark committed:
Weakref info button

LemonInTheDark committed:
Fixes a plant gene harddel, pass by copy not by ref bud

LemonInTheDark committed:
Moves many ai behaviors over to weakrefs, with the goal of preventing the (set var, never unset, assume it will still exist) pattern. Fixes monkey sourced target harddels, so humans and slimes mostly

LemonInTheDark committed:
Converts a pictures human contents lists to weakrefs
God the alagory is holding strong, taking photos of someone really does steal their soul

LemonInTheDark committed:
Suppose it should compile

LemonInTheDark committed:
Cleans up surgery init a bit, fixes a mob/limb harddel, surgeries weren't cleaning up on destroy

LemonInTheDark committed:
Fixes beartrap harddels. They were being removed, qdeling, losing their flag, then dropping. The flag removal existed to prevent loops, so let's just use QDELETED instead

LemonInTheDark committed:
Small cleanups

LemonInTheDark committed:
Adds logging for attempting to give a mob a non string name. Hopefully this helps to catch the code that's setting a mob's real_name to an observer mob object

LemonInTheDark committed:
QDELETED is better here. I know something is causing the construct's cell var to hang a ref, but sure what tho

LemonInTheDark committed:
Oh come on

LemonInTheDark committed:
Fixes an action that holds its parent hhhhhhhhh

LemonInTheDark committed:
Ok so now the monitor itself is being held

LemonInTheDark committed:
Jesus christ that's a lot of actions

LemonInTheDark committed:
Fixes a manifest harddel, New -> Initialize, fixes a potential qdel in supply ordering caused by a manifest paper being inside of contents before contents has a part of it qdel'd

LemonInTheDark committed:
Fixes cotten trying to merge with qdeleted things

LemonInTheDark committed:
Sadkirby

LemonInTheDark committed:
Ai var/Bot to weakrefs, jesus

LemonInTheDark committed:
Cleans up medigun/current_beam properly

LemonInTheDark committed:
Merge branch 'here-we-go' into here-we-suffer

LemonInTheDark committed:
Makes it compile

LemonInTheDark committed:
Merge branch 'here-we-go' into here-we-suffer

LemonInTheDark committed:
Fixes a growth_queue harddelete, hardens the spacevine controller a bit and cleans up er code

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into here-we-go

LemonInTheDark committed:
Fixes a regression with important_recursive_contents, got moved up in an attempt to fix a bug that kyler assures me was fixed in a different way. Here's hoping

LemonInTheDark committed:
Fixes a modular computer harddel, turns out a harddrive cam be removed without its computer deleting, which causes headaches when clearing up refs

LemonInTheDark committed:
Cleans up modular computer programs a bit, fixing ref hangs, cleaning up useless vars, etc

LemonInTheDark committed:
Cleans up kiosks, fixes a few harddels. Jesus fucking christ

LemonInTheDark committed:
Converts the monkey enemies list to weakrefs

LemonInTheDark committed:
Converts monkey enemies to weakrefs

LemonInTheDark committed:
Merge branch 'here-we-go' into here-we-suffer

LemonInTheDark committed:
Wahhhh

LemonInTheDark committed:
Merge branch 'here-we-go' into here-we-suffer

LemonInTheDark committed:
wow your mom lets you compile AND run?

LemonInTheDark committed:
Merge branch 'here-we-go' into here-we-suffer

LemonInTheDark committed:
Son I am fucking disappoint

LemonInTheDark committed:
Merge branch 'here-we-go' into here-we-suffer

LemonInTheDark committed:
Fixes client cleanup related harddels. Basically we need a ref to the client in mob destroy, so we manage our own. Also fixes up some observe runtimes

LemonInTheDark committed:
Merge branch 'here-we-go' into here-we-suffer

LemonInTheDark committed:
Hopefully fixes a tackle component sourced harddel, we don't really own the tackle, so let's weakref it

LemonInTheDark committed:
Removes a weird python like typebuilding pattern with a proper fullscreen subtype, potentially fixes some weird /obj harddels I've been seeing recently, we'll have to see

LemonInTheDark committed:
Kills the other two instances of /obj instanciation I can find, if we're gonna do this awful hack we should at least mark it as one. Also cleans up the code a smidge

LemonInTheDark committed:
Lightly firebombs expirementor code, fixing a pair of ian and runtime harddels

LemonInTheDark committed:
Fixes an inserted_key harddel in vehicle code in 2 different ways. Is it just me or are trailers unused

LemonInTheDark committed:
Fixes a random manuel spawner harddel, it dels in init. Newing something does not somehow magically return the object that was spawned in as a side effect of the new you fucking reeeeeeeeeeeee

LemonInTheDark committed:
Fixes some stupid code that hangs refs in the conjure spell, alongside not properly working for all potential cases

LemonInTheDark committed:
Fixes assembly holders hanging refs to the assemblies inserted into them. Sadge

LemonInTheDark committed:
Fixes flamethrower sourced harddels, QDEL_NULL instead of qdel you hhhhhhhhhh

LemonInTheDark committed:
Fixes onetankbomb sourced harddels, god these things are just horrible

LemonInTheDark committed:
Makes the master var actually garbage collect, and marks it as deprecated. Should honestly just remove it, but it's late. Prehaps tomorrow

LemonInTheDark committed:
Merge branch 'here-we-go' into here-we-suffer

LemonInTheDark committed:
Fixes some cartridge sourced bot harddels

LemonInTheDark committed:
Makes the team_hud var a weakref. Not sure why it wasn't one before, since its type can sometimes self delete

LemonInTheDark committed:
Ok this is gonna be a big one

I'm redoing how camera consoles handle actions
Instead of storing all active actions in the actions list, inserting them from external vars on mob interact
We insert any actions we want our mob to have into the actions list
Then GrantActions handles the rest.

As a part of this, I've moved some uses of the target var to owner, so target can be properly set in action
init.

I'm also making the abductor console and its tools properly gc

LemonInTheDark committed:
Merge branch 'here-we-go' into here-we-suffer

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into here-we-go

LemonInTheDark committed:
Merge branch 'here-we-go' into here-we-suffer

LemonInTheDark committed:
Parently code needs to compile. Who knew

LemonInTheDark committed:
Merge branch 'here-we-go' into here-we-suffer

LemonInTheDark commented:
Calling it here, was a good run. Will likely do more in future, but it'll be easier for others to do what I do without needing to be a maintainer/needing indepth knowledge of reftracking
